### PR TITLE
[BEAM-10166] Improve execution time errors with more concise error messages from DoFns

### PR DESCRIPTION
**Pull request by codeBehindMe**
The Go SDK uses errors returned by DoFns to signal failures to process bundles, and terminate bundle processing. However, if the preceding DoFn uses emitters, rather than error returns, the code has no choice to panic to avoid user code handling or ignoring the cross DoFn error (which could cause dataloss or other correctness problems).

All bundle executions are wrapped in `callNoPanic` to prevent worker termination on such panics, and orderly terminate just the affected bundle instead.`callNoPanic` uses Go's built in recover mechanism to get the error and provide a stack trace.

This is message is not necessarily helpful to the user and can lead to degraded user experience when debugging such errors.

In this patch, we introduce a new type of error; a `doFnError` which has a more concise representation of the DoFn which failed, the error which the DoFn failed with and other meta information.

------------------------

Thank you for your contribution! Follow this checklist to help us incorporate your contribution quickly and easily:

 - [x] [**Choose reviewer(s)**](https://beam.apache.org/contribute/#make-your-change) and mention them in a comment (`R: @username`).
 - [x] Format the pull request title like `[BEAM-XXX] Fixes bug in ApproximateQuantiles`, where you replace `BEAM-XXX` with the appropriate JIRA issue, if applicable. This will automatically link the pull request to the issue.
 - [x] Update `CHANGES.md` with noteworthy changes.
 - [ ] If this contribution is large, please file an Apache [Individual Contributor License Agreement](https://www.apache.org/licenses/icla.pdf).

See the [Contributor Guide](https://beam.apache.org/contribute) for more tips on [how to make review process smoother](https://beam.apache.org/contribute/#make-reviewers-job-easier).

Post-Commit Tests Status (on master branch)
------------------------------------------------------------------------------------------------

Lang | SDK | Dataflow | Flink | Samza | Spark | Twister2
--- | --- | --- | --- | --- | --- | ---
Go | [Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Go/lastCompletedBuild/) | --- | [Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Go_VR_Flink/lastCompletedBuild/) | --- | [Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Go_VR_Spark/lastCompletedBuild/) | ---
Java | [Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Java/lastCompletedBuild/) | [Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Dataflow/lastCompletedBuild/)<br>[Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Dataflow_Java11/lastCompletedBuild/) | [Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Flink/lastCompletedBuild/)<br>[Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Flink_Java11/lastCompletedBuild/)<br>[Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Java_PVR_Flink_Batch/lastCompletedBuild/)<br>[Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Java_PVR_Flink_Streaming/lastCompletedBuild/) | [Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Samza/lastCompletedBuild/) | [Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Spark/lastCompletedBuild/)<br>[Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Java_PVR_Spark_Batch/lastCompletedBuild/)<br>[Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Java_ValidatesRunner_SparkStructuredStreaming/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Twister2/lastCompletedBuild/)
Python | [Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Python2/lastCompletedBuild/)<br>[Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Python35/lastCompletedBuild/)<br>[Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Python36/lastCompletedBuild/)<br>[Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Python37/lastCompletedBuild/) | [Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Py_VR_Dataflow/lastCompletedBuild/)<br>[Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Py_VR_Dataflow_V2/lastCompletedBuild/)<br>[Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Py_ValCont/lastCompletedBuild/) | [Build Status](https://ci-beam.apache.org/job/beam_PreCommit_Python2_PVR_Flink_Cron/lastCompletedBuild/)<br>[Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Python35_VR_Flink/lastCompletedBuild/) | --- | [Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Python_VR_Spark/lastCompletedBuild/) | ---
XLang | --- | --- | [Build Status](https://ci-beam.apache.org/job/beam_PostCommit_XVR_Flink/lastCompletedBuild/) | --- | --- | [Build Status](https://ci-beam.apache.org/job/beam_PostCommit_XVR_Spark/lastCompletedBuild/) | ---

Pre-Commit Tests Status (on master branch)
------------------------------------------------------------------------------------------------

--- |Java | Python | Go | Website
--- | --- | --- | --- | ---
Non-portable | [Build Status](https://ci-beam.apache.org/job/beam_PreCommit_Java_Cron/lastCompletedBuild/) | [Build Status](https://ci-beam.apache.org/job/beam_PreCommit_Python_Cron/lastCompletedBuild/)<br>[Build Status](https://ci-beam.apache.org/job/beam_PreCommit_PythonLint_Cron/lastCompletedBuild/) | [Build Status](https://ci-beam.apache.org/job/beam_PreCommit_Go_Cron/lastCompletedBuild/) | [Build Status](https://ci-beam.apache.org/job/beam_PreCommit_Website_Cron/lastCompletedBuild/)
Portable | --- | [Build Status](https://ci-beam.apache.org/job/beam_PreCommit_Portable_Python_Cron/lastCompletedBuild/) | --- | ---

See [.test-infra/jenkins/README](https://github.com/apache/beam/blob/master/.test-infra/jenkins/README.md) for trigger phrase, status and link of all Jenkins jobs.

### CODE REVIEW

**Comment by codeBehindMe** [comment]
R: @lostluck

File: sdks/go/pkg/beam/core/runtime/exec/util.go [hunk 1]
```diff
@@ -27,6 +28,28 @@ type GenID struct {
...
+// caught where the origin is due to a failed DoFn execution in process bundle
+// execution.
+type doFnError struct {
+	doFn string
+	err  error
+	uid  UnitID
+	pid  string
+}
+func (e *doFnError) Error() string {
+	return fmt.Sprintf("%v caused error %v, uid is %v pid is %v", e.doFn, e.err, e.uid, e.pid)
```
Comment (lostluck) [inline sdks/go/pkg/beam/core/runtime/exec/util.go:50]: The wiki syntax in JIRA really butchered my intended suggestion of `DoFn[<uid>;<pid>]<doFn> returned error:<error>`

Which places the  short/known length things before the potentially long and rife with newlines error. It puts the context first, and avoids it getting lost or truncated behind a very long error string.

**Comment by lostluck** [comment]
This is a good first step! Next up is adding it for use in user error returns in exec/pardo.go and similar.

If you did that work already, it's probable that the commit wasn't pushed to your branch.

Thanks for your patience in my getting to this. I took a long weekend for the US (and Canadian) holiday, and avoided all pretense of work.

File: sdks/go/pkg/beam/core/runtime/exec/util.go (hunk 1 above)
Comment (codeBehindMe) [reply sdks/go/pkg/beam/core/runtime/exec/util.go]: This is an excellent suggestion.
ff46c6a783

**Comment by lostluck** [comment]
This looks good to me! Thank you for the contribution!

Would you mind adding a test to validate this behaviour?
exec/util_test.go doesn't currently exist, so it would need to be added.

There's no need to test everything in exec/util.go, just callNoPanic.  

You can test three cases: 
*  one where the passed in func(context.Context) error doesn't panic, and the error is just returned as is (they should compare as equal for a simple error)
 * another test where the the passed in function panics, which should return an error that contains "panic: " in it. 
  * finally, one where the error passed to panic is wrapped in your new doFnError error type.

**Comment by codeBehindMe** [comment]
thank @lostluck , I will implement this today / tomorrow. 
Apologies, personal commitments have been getting in the way recently.

**Comment by lostluck** [comment]
Definitely no need to apologize. Life >> code.

**Comment by codeBehindMe** [comment]
So i've been banging my head against the wall with the following error. Am I missing something obvious here?
image

**Comment by lostluck** [comment]
Your image is missing the whole error, but I suspect that the build doesn't know where that package is coming from.

`go get github.cmo/google/go-cmp/cmp/cmpopts` should solve the problem locally (and repeat until complete).

An alternative that should work is if you go to the beam root (.../github.com/apache/beam/) and run the gogradle vendor command.  `./gradlew :sdks:go:goVendor`  This requires java and such to be installed unfortunately IIRC.  This does almost the same as the above but puts everything into a vendor directory so it only impacts the Go SDK.

Unfortunately I can't recommend the best solution since it's not guaranteed to work with the way the Go SDK is currently implemented, which should get everything for you in one command. There's a small bit of logistics that need to be accomplished and feature work finished before we fix that though.

**Comment by codeBehindMe** [comment]
Thanks @lostluck let me try my luck! =D

**Pull request closed without merging**